File: framework/languages/java/e2e/ResilienceLifecycle/Shared/src/main/java/systems/zlink/e2e/resiliencelifecycle/shared/Contracts.java
package systems.zlink.e2e.resiliencelifecycle.shared;

import java.util.List;

public final class Contracts {
    public static final String CHANNEL = "resilience.lifecycle.api";
    public static final String HANDLER_GROUP = "resilience-lifecycle";

    private Contracts() {
    }

    public record WorkReq(String value) {
    }

    public record WorkMsg(String value) {
    }

    public record UnhandledReq(String value) {
    }

    public record WorkRes(
        String value,
        String providerRid) {
    }

    public record WorkOperation(String value, long timeoutMillis) {
    }

    public record UnhandledOperation(String value, long timeoutMillis) {
    }

    public record PeerLocation(
        String routingId,
        String endpoint,
        String ownerId,
        long generation) {
    }

    public record PeerSnapshot(List<PeerLocation> peers) {
    }

    public record EvidenceEntry(
        String marker,
        String providerRid,
        String value) {
    }

    public record EvidenceSnapshot(
        String providerRid,
        int weight,
        List<EvidenceEntry> entries) {
    }
}

File: framework/languages/java/e2e-kotlin/ResilienceLifecycle/Shared/src/main/kotlin/systems/zlink/e2e/kotlin/resiliencelifecycle/Contracts.kt
package systems.zlink.e2e.kotlin.resiliencelifecycle

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty

object Contracts {
    const val CHANNEL: String = "resilience.lifecycle.api"
    const val HANDLER_GROUP: String = "resilience-lifecycle"

    data class WorkReq @JsonCreator constructor(
        @JsonProperty("value") val value: String,
    ) {
        fun value(): String = value
    }

    data class WorkMsg @JsonCreator constructor(
        @JsonProperty("value") val value: String,
    ) {
        fun value(): String = value
    }

    data class UnhandledReq @JsonCreator constructor(
        @JsonProperty("value") val value: String,
    ) {
        fun value(): String = value
    }

    data class WorkRes @JsonCreator constructor(
        @JsonProperty("value") val value: String,
        @JsonProperty("providerRid") val providerRid: String,
    ) {
        fun value(): String = value

        fun providerRid(): String = providerRid
    }

    data class EvidenceEntry @JsonCreator constructor(
        @JsonProperty("marker") val marker: String,
        @JsonProperty("providerRid") val providerRid: String,
        @JsonProperty("value") val value: String,
    ) {
        fun marker(): String = marker

        fun providerRid(): String = providerRid

        fun value(): String = value
    }

    data class EvidenceSnapshot @JsonCreator constructor(
        @JsonProperty("providerRid") val providerRid: String,
        @JsonProperty("weight") val weight: Int,
        @JsonProperty("entries") val entries: List<EvidenceEntry>,
    ) {
        fun providerRid(): String = providerRid

        fun weight(): Int = weight

        fun entries(): List<EvidenceEntry> = entries
    }

    data class TopologyWaitReq @JsonCreator constructor(
        @JsonProperty("expectedRouters") val expectedRouters: Int,
        @JsonProperty("routingId") val routingId: String?,
        @JsonProperty("endpoint") val endpoint: String?,
    )

    data class TopologyWaitRes @JsonCreator constructor(
        @JsonProperty("matchedRouters") val matchedRouters: Int,
    )

    data class TopologyReadRes @JsonCreator constructor(
        @JsonProperty("status") val status: String,
        @JsonProperty("matchedRouters") val matchedRouters: Int,
        @JsonProperty("error") val error: String?,
    )
}
